pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskExecutor.java

```diff
@@ -0,0 +1,183 @@
+        values.add(row.getValue(column).toString());
+      }
+      return !validIds.contains(values.hashCode());
+    });
+    SegmentPurger.RecordPurger recordPurger = MINION_CONTEXT.getRecordPurgerFactory().getRecordPurger(rawTableName);
+
+    _eventObserver.notifyProgress(_pinotTaskConfig, "Generating segment");
+    SegmentPurger segmentPurger = new SegmentPurger(indexDir, workingDir, tableConfig, recordPurger, null);
+    File compactedSegmentFile = segmentPurger.purgeSegment();
+    if (compactedSegmentFile == null) {
+      compactedSegmentFile = indexDir;
+    }
+
+    SegmentConversionResult result = new SegmentConversionResult.Builder()
+        .setFile(compactedSegmentFile)
+        .setTableNameWithType(tableNameWithType)
+        .setSegmentName(configs.get(MinionConstants.SEGMENT_NAME_KEY))
+        .setCustomProperty(RECORD_PURGER_KEY, segmentPurger.getRecordPurger())
+        .setCustomProperty(NUM_RECORDS_PURGED_KEY, segmentPurger.getNumRecordsPurged()).build();
+
+    long endMillis = System.currentTimeMillis();
+    LOGGER.info("Finished task: {} with configs: {}. Total time: {}ms", taskType, configs, (endMillis - startMillis));
+
+    return result;
+  }
+
+  private static ImmutableRoaringBitmap getValidDocIds(String tableNameWithType, Map<String, String> configs) {
+    String segmentName = configs.get(MinionConstants.SEGMENT_NAME_KEY);
+    String server = getServer(segmentName, tableNameWithType);
+
+    // get the url for the validDocIds for the server
+    InstanceConfig instanceConfig =
+        MINION_CONTEXT.getClusterManagementTool().getInstanceConfig(MINION_CONTEXT.getClusterName(), server);
+    String endpoint = InstanceUtils.getServerAdminEndpoint(instanceConfig);
+    String url = String.format("%s/segments/%s/%s/validDocIds",
+        endpoint, tableNameWithType, segmentName);
+
+    // get the validDocIds from that server
+    Response response = ClientBuilder.newClient().target(url).request().get(Response.class);
+    Preconditions.checkState(response.getStatus() == Response.Status.OK.getStatusCode(),
+        "Unable to retrieve validDocIds from %s", url);
+    byte[] snapshot = response.readEntity(byte[].class);
+    ImmutableRoaringBitmap validDocIds = new ImmutableRoaringBitmap(ByteBuffer.wrap(snapshot));
+    return validDocIds;
+  }
+
+  @VisibleForTesting
+  public static Set<Integer> getValidIds(ImmutableRoaringBitmap validDocIds,
+      File indexDir, List<String> columns) throws IOException {
+    Set<Integer> validIds = new HashSet<>();
+
+    if (validDocIds.isEmpty()) {
+      return validIds;
+    }
+
+    PeekableIntIterator iterator = validDocIds.getIntIterator();
+    PinotSegmentRecordReader recordReader = new PinotSegmentRecordReader();
+    recordReader.init(indexDir, new HashSet<>(columns), null);
+    GenericRow genericRow = new GenericRow();
+
+    while (iterator.hasNext()) {
+      int validDocId = iterator.next();
+      recordReader.getRecord(validDocId, genericRow);
```

Comment:
Note: This would do a lot of Random IO.

Reply:
@ankitsultana Would this be better?
```
int docId = 0;
while (recordReader.hasNext()) {
  recordReader.next(genericRow);
  if (validDocIds.contains(docId)) {
    validIds.add(genericRow.getPrimaryKey(columns).hashCode());
  }
  docId++;
}
```

Reply:
The alternative you mentioned will also do the same amount of random IO (the issue is within `recordReader.next(genericRow)`).

I don't think this is something to worry about too much since the data being compacted shouldn't be super high in one batch, and in most cases users may be running minions separate from pinot servers and brokers.

But for users running things at scale this is something to be aware of. The fix for this would be a bit complex to implement (read each column 1 by 1 and create rows in memory).